### PR TITLE
Relax ordering constraints for parameter modifiers

### PR DESCRIPTION
**Pull request by alrz**
### Customer scenario
We allow `ref this` in ref extension methods but not the other way around. That's wrong! We can't disallow `ref this`, but we should allow and prefer `this ref`.

Note: this is a bug fix change to C# 7.2 that would ship in 15.6, per discussion with LDM.

### Bugs this fixes

Fixes https://github.com/dotnet/csharplang/issues/1022 (per LDM [2017-12-04](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-12-04.md))

### Workarounds, if any
Type `ref this` instead of `this ref`.

### Performance impact
TBD

### Is this a regression from a previous update?
No. This restriction was introduced when the ref extension method was implemented in C# 7.2 (15.5).

### How was the bug found?
Reported by customer.

### CODE REVIEW

**Comment by jcouv** [comment]
Tagging @VSadov as FYI. Although we hadn't discussed `in` (only `ref`), the same relaxation on ordering makes sense for `in` too (as this PR implements).

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 1]
```diff
@@ -1517,7 +1517,7 @@ internal enum ErrorCode
         ERR_TypeReserved = 8336,
         ERR_RefExtensionMustBeValueTypeOrConstrainedToOne = 8337,
         ERR_InExtensionMustBeValueType = 8338,
-        ERR_BadParameterModifiersOrder = 8339,
+        // ERR_BadParameterModifiersOrder = 8339,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: >// ERR_BadParameterModifiersOrder = 8339, [](start = 8, length = 41)

I would expect the error message to be removed from the .resx file

**Comment by AlekseyTs** [comment]
I would expect to see some tests that verify that the different order of modifiers is actually properly treated by the compiler, i.e. the parameter is actually treated as ref/in by the compiler in addition to not reporting the error.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1).

**Comment by VSadov** [comment]
Yes, it applies equally to ‘ref’ and ‘in’ extensions.

**Comment by VSadov** [comment]
Also need some tests for scenarios with possibly new behavior like:  
“ref this ref”, 
“ref this in”

Still an error, but for a different reason.

**Comment by alrz** [comment]
@AlekseyTs @VSadov Ready for review.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs (hunk 1 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: I would also add a small comment to explain why it was removed.

---
In reply to: [156157236](https://github.com/dotnet/roslyn/pull/23643#discussion_r156157236) [](ancestors = 156157236)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 2]
```diff
@@ -3524,16 +3518,7 @@ public static class TestType
...
-                // public static void Method6<T, U, V>(this in int i) { }
-                Diagnostic(ErrorCode.ERR_BadParameterModifiersOrder, "in").WithArguments("in", "this").WithLocation(14, 42),
-                // (6,33): error CS8339: The parameter modifier 'in' cannot be used after the modifier 'this'
-                // public static void Method2(this in int i) { }
-                Diagnostic(ErrorCode.ERR_BadParameterModifiersOrder, "in").WithArguments("in", "this").WithLocation(6, 33),
-                // (10,36): error CS8339: The parameter modifier 'in' cannot be used after the modifier 'this'
-                // public static void Method4<T>(this in int i) { }
-                Diagnostic(ErrorCode.ERR_BadParameterModifiersOrder, "in").WithArguments("in", "this").WithLocation(10, 36));
+            CreateCompilationWithMscorlibAndSystemCore(test).GetDeclarationDiagnostics().Verify();
         }
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: I'd use `Compilation.VerifyDiagnostics()` for completeness.

Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: I'd also change the name of the test.

---
In reply to: [157593732](https://github.com/dotnet/roslyn/pull/23643#discussion_r157593732) [](ancestors = 157593732)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 3]
```diff
@@ -3112,53 +3112,47 @@ public static class GenExtensions<X> where X : struct
...
-                //     public static void Goo2(this ref int i) {}
-                Diagnostic(ErrorCode.ERR_BadParameterModifiersOrder, "ref").WithArguments("ref", "this").WithLocation(15, 34),
                 // (12,21): error CS1106: Extension method must be defined in a non-generic static class
-                // public static class GenExtensions<X>
+                // public static class GenExtensions<X> where X : struct
                 Diagnostic(ErrorCode.ERR_BadExtensionAgg, "GenExtensions").WithLocation(12, 21));
         }
 
+        [Fact]
+        public void BadParameterModifiers()
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: >BadParameterModifiers [](start = 20, length = 21)

Can we change the test name to describe which modifiers are bad?

Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: I suggest adding the opposite test scenarios as well:

* Testing we produce an error on `in this ref int i`
* Testing we produce an error on `in this in int i`

---
In reply to: [157594285](https://github.com/dotnet/roslyn/pull/23643#discussion_r157594285) [](ancestors = 157594285)

**Comment by OmarTawfik** [comment]
>         public void BadParameterModifiers_ThisWithRef()

Can we rename that test to be more descriptive? this is now testing producing errors on the containing type.

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs:3087 in f4a4534. [](commit_id = f4a4534e1a16797f1060bc35e9821361a01dbdec, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs [hunk 4]
```diff
@@ -1965,6 +1965,26 @@ public static class Ext
             Assert.Equal(RefKind.In, symbol.RefKind);
         }
 
+        [Fact]
+        public void ParameterSymbolsRetrievedThroughSemanticModel_ModifierOrdering()
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs]: >ParameterSymbolsRetrievedThroughSemanticModel_ModifierOrdering [](start = 20, length = 62)

This title is misleading. The symbols are not retrieved through the semantic model. I suggest following what `ParameterSymbolsRetrievedThroughSemanticModel_InExtensionMethods` do. I also suggest separating `this in` and `this ref`.

Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs]: That test is still useful, but I suggest moving it to the Symbols test project, and use `CompileAndVerify()` helper, making use of the `symbolValidator` parameter to check for the symbols.

---
In reply to: [157595248](https://github.com/dotnet/roslyn/pull/23643#discussion_r157595248) [](ancestors = 157595248)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 5]
```diff
@@ -3505,7 +3458,7 @@ public class TestType
 
         [Fact]
         [CompilerTrait(CompilerFeature.ReadOnlyReferences)]
-        public void BadInWithThisParameterModifiers()
+        public void InWithThis_ParameterModifiers()
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: We don't need this test now. VerifyDiagnostics is called on the same combination of modifiers elsewhere.
But if we do want to keep this, it should be moved to another file.

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs [hunk 6]
```diff
@@ -3882,5 +3882,65 @@ public class BaseClass<TMember> : I1<TMember>
...
+
+                method = type.GetMember<MethodSymbol>("M2");
+                Assert.True(method.IsExtensionMethod);
+                parameter = method.Parameters[0];
+                Assert.Equal(parameter.Type.SpecialType, SpecialType.System_Int32);
+                Assert.Equal(RefKind.In, parameter.RefKind);
+            }
+
+            CompileAndVerify(source, validator: Validator, options: TestOptions.ReleaseDll).VerifyDiagnostics();
+        }
```
Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs]: nit: for these two methods, `CompileAndVerify` will throw if there are any errors, so the trailing `VerifyDiagnostics()` is not needed.
However, you might want to make these tests actually print something through the extension methods, and use `expectedOutput: "XX"` parameter to verify the result.

**Comment by jcouv** [comment]
@MeiChin-Tsai for ask-mode approval for 15.6. Thanks

**Comment by MeiChin-Tsai** [comment]
approved.

**Pull request merged**